File: CastleEscape/TextMenu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.GamerServices;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using Microsoft.Xna.Framework.Media;
using Microsoft.Xna.Framework.Net;
using Microsoft.Xna.Framework.Storage;

namespace CastleEscape
{
    /// <summary>
    /// Allows the user to select from a list of options.
    /// 
    /// Author: Dennis Honeyman
    /// </summary>
    class TextMenu
    {
        private SpriteFont font;
        private string[] options;
        private int selectedOption = 0;
        private float defaultSpacing;
        private float selectedSpacing;
        private bool canMove;
        private bool isFinished;
        private bool canPressZ;

        public bool IsFinished
        {
            get { return isFinished; }
            set { isFinished = value; }
        }

        public int SelectedOption
        {
            get { return selectedOption; }
        }

        public string[] Options
        {
            get { return options; }
            set { options = value; }
        }

        public TextMenu(SpriteFont font, string[] options)
        {
            this.options = options;
            this.font = font;
            defaultSpacing = font.Spacing;
            canMove = false;
            isFinished = false;
            canPressZ = false;
        }

        public void Update(GameTime gameTime, KeyboardState state)
        {
            if (!isFinished)
            {
                selectedSpacing = (float)(defaultSpacing + Math.Sin(Math.Log(gameTime.TotalGameTime.Milliseconds)) * 20);
                selectedSpacing = selectedSpacing / options[selectedOption].Length;
            }
            else
            {
                canMove = false;
                selectedSpacing = defaultSpacing;
                return;
            }

            if (state.IsKeyUp(Keys.Z))
                canPressZ = true;

            if (canPressZ && state.IsKeyDown(Keys.Z))
            {
                isFinished = true;
                canPressZ = false;
                selectedSpacing = defaultSpacing;
            }

            if (state.IsKeyUp(Keys.Up) && state.IsKeyUp(Keys.Down))
                canMove = true;
            if (!canMove)
                return;
            if (state.IsKeyDown(Keys.Up))
            {
                selectedOption = selectedOption > 0 ? selectedOption - 1 : options.Length - 1;
                canMove = false;
            }
            else if (state.IsKeyDown(Keys.Down))
            {
                selectedOption = (selectedOption + 1) % options.Length;
                canMove = false;
            }
        }

        public void Draw(SpriteBatch spriteBatch, int x, int y)
        {
            for (int i = 0; i < options.Length; i++)
            {
                var pos = new Vector2(x, y + i * font.LineSpacing);
                if (i == selectedOption && !isFinished)
                {
                    pos.X += 2 * selectedSpacing;
                    font.Spacing = selectedSpacing;
                }
                else
                    font.Spacing = defaultSpacing;
                spriteBatch.DrawString(font, options[i], pos, Color.Black);
            }
            font.Spacing = defaultSpacing;
        }
    }
}

File: Source/TextMenu.cc
#include "TextMenu.hh"

#include <cmath>
#include <string>
#include <vector>

#include <SFML/Graphics.hpp>

namespace CastleEscape {

TextMenu::TextMenu(const std::vector<std::string>& options,
		const sf::Font& font) {
	options_ = options;
	font_ = font;
	selectedOption_ = 0;
	defaultStretch_ = 1;
	canMove_ = false;
	isFinished_ = false;
	canPressZ_ = false;
}

bool TextMenu::IsFinished() {
	return isFinished_;
}

void TextMenu::Restart() {
	isFinished_ = false;
}

uint TextMenu::GetSelectedOption() {
	return selectedOption_;
}

void TextMenu::Update(const sf::Clock& clock, const sf::Input& input) {
	if (!isFinished_) {
		selectedStretch_ = (float) (defaultStretch_ + std::sin(
				clock.GetElapsedTime() * 5));
		selectedStretch_ = 1 + selectedStretch_ / options_[selectedOption_].size();
	} else {
		canMove_ = false;
		selectedStretch_ = defaultStretch_;
		return;
	}

	if (!input.IsKeyDown(sf::Key::Z))
		canPressZ_ = true;

	if (canPressZ_ && input.IsKeyDown(sf::Key::Z)) {
		isFinished_ = true;
		canPressZ_ = false;
		selectedStretch_ = defaultStretch_;
	}

	if (!input.IsKeyDown(sf::Key::Up) && !input.IsKeyDown(sf::Key::Down))
		canMove_ = true;
	if (!canMove_)
		return;
	if (input.IsKeyDown(sf::Key::Up)) {
		selectedOption_ = selectedOption_ > 0 ? selectedOption_ - 1
				: options_.size() - 1;
		canMove_ = false;
	} else if (input.IsKeyDown(sf::Key::Down)) {
		selectedOption_ = (selectedOption_ + 1) % options_.size();
		canMove_ = false;
	}
}

void TextMenu::Draw(sf::RenderWindow& window, int x, int y,
		const sf::Color& color) {
	for (uint i = 0; i < options_.size(); i++) {
		uint charSize = font_.GetCharacterSize();
		sf::String item(options_[i], font_, charSize);
		item.SetPosition(x, y + i * charSize);
		item.SetColor(color);
		if (i == selectedOption_)
			item.SetScaleX(selectedStretch_);
		window.Draw(item);
	}
}

} // namespace CastleEscape
